agent/proxycfg/state.go

```diff
@@ -616,13 +621,26 @@ func (s *state) run() {
 			// etc on future updates.
 			snapCopy, err := snap.Clone()
 			if err != nil {
-				s.logger.Error("Failed to copy config snapshot for proxy",
-					"proxy", s.proxyID,
-					"error", err,
+				logger.Error("Failed to copy config snapshot for proxy",
+					"proxy-id", s.proxyID.String(), "error", err,
 				)
 				continue
 			}
-			s.snapCh <- *snapCopy
+
+			select {
+			// try to send
+			case s.snapCh <- *snapCopy:
+				logger.Trace("Delivered new snapshot to proxy config watchers",
+					"proxy-id", s.proxyID.String(),
+				)
+
+			// avoid blocking if a snapshot is already buffered
+			default:
+				logger.Trace("Failed to deliver new snapshot to proxy config watchers",
+					"proxy-id", s.proxyID.String(),
+				)
```

Comment:
I see how this breaks the deadlock, but is it OK? If we just drop the new snapshot delivery on the floor that means that some change in proxy config that happened since the chan was last filled is now never going to get applied to any proxies consuming this state (at least until something else changes).

Isn't that just making the deadlock into a silent (other than the trace log) delivery failure that won't automatically resolve?

I could understand if the chan were just a "something changed" edge signal for the proxy to reload to the current config state, but it's not it's delivering a concrete state and now if we do hit this race, we'll just drop some valid config change on the floor won't we?

Reply:
Would it be correct if we instead of dropping the update entirely, just scheduled a new deliver of it later (by sending on the sendCh again.

That would unblock us here, release the deadlock, but then in a few milliseconds we'd be back and would finally get another chance to deliver the latest snapshot instead of dropping it?

Reply:
We're already scheduling a new delivery of it though. 

1. After dropping the send, [here](https://github.com/hashicorp/consul/blob/proxycfg-deadlock/agent/proxycfg/state.go#L645) we're resetting the coalesceTimer 
2. In a following loop we fall into [this block](https://github.com/hashicorp/consul/blob/proxycfg-deadlock/agent/proxycfg/state.go#L683) where we re-create the coalesceTimer since the snapshot is valid
3. After the timer elapses (200ms), we end up [in the case where we started](https://github.com/hashicorp/consul/blob/proxycfg-deadlock/agent/proxycfg/state.go#L619)

So then a question is: is 200ms too long?

Reply:
Hmm I'm not sure if I follow.

We reset the timer, but then we `continue` which means we skip the code at the bottom of the loop and start waiting for a further change. Doesn't that mean we will just wait without sending the actual update until _something else_ triggers a change? That could be an unbounded amount of time later.

Reply:
Ah yep I mistakenly thought we would fall through the select statement in a later loop even if nothing happened.